core/commonMain/src/kotlinx/serialization/internal/Enums.kt

```diff
@@ -49,20 +50,54 @@ internal class EnumDescriptor(
     }
 }
 
-// Used for enums that are not explicitly serializable by the plugin
+@OptIn(ExperimentalSerializationApi::class)
+@InternalSerializationApi
+internal fun <T : Enum<T>> createSimpleEnumSerializer(serialName: String, values: Array<T>): KSerializer<T> {
+    return EnumSerializer(serialName, values)
+}
+
+@OptIn(ExperimentalSerializationApi::class)
+@InternalSerializationApi
+internal fun <T : Enum<T>> createMarkedEnumSerializer(
+    serialName: String,
+    values: Array<T>,
+    names: Array<String?>,
+    annotations: Array<Array<Annotation>?>
+): KSerializer<T> {
+    val descriptor = EnumDescriptor(serialName, values.size)
+    values.forEachIndexed { i, v ->
+        val elementName = names.getOrNull(i) ?: v.name
+        descriptor.addElement(elementName)
+        annotations.getOrNull(i)?.forEach {
+            descriptor.pushAnnotation(it)
+        }
+    }
+
+    return EnumSerializer(serialName, values, descriptor)
+}
+
+// TODO we can create another class for factories, it may be a copy-paste of this class or a subclass for some `AbstractEnumSerializer` with common `serialize`, `deserialize` and `toString` functions
 @PublishedApi
 @OptIn(ExperimentalSerializationApi::class)
 internal class EnumSerializer<T : Enum<T>>(
     serialName: String,
     private val values: Array<T>
 ) : KSerializer<T> {
+    @Volatile
```

Comment:
Is `@Volatile` really needed here? I don't see why

Reply:
Indeed, there is no need for it in the current implementation